Added TimeLock Manger Reversible Contract with the Tests

Ashar2shahid opened the pull request.

bbenligiray reviewed (changes requested):
I couldn't check the tests yet. Please also set up a check for test coverage, see this for reference:
https://github.com/api3dao/airnode/tree/master/packages/protocol

Ashar2shahid commented on packages/timelock-manager/contracts/TimeLockManagerReversible.sol:
```diff
@@ -0,0 +1,260 @@
+import "@api3-contracts/api3-token/contracts/interfaces/IApi3Token.sol";
+import "./interfaces/ITimelockManagerReversible.sol";
+
+/// @title Contract that the API3 DAO uses to timelock API3 tokens
```
All right will do

Ashar2shahid replied:
I ran the test coverages, I've adjusted the tests for it to pick up stopVesting. Some of the statements that are also present in the original `timelock-manager` are not being covered like:
```
function withdraw()
        external
        override
        onlyIfRecipientHasRemainingTokens(msg.sender)
    {
        address recipient = msg.sender;
        uint256 withdrawable = getWithdrawable(recipient);
        Irequire(withdrawable != 0, "No withdrawable tokens yet");
        timelocks[recipient].remainingAmount = timelocks[recipient]
            .remainingAmount
            .sub(withdrawable);
        require(
            api3Token.transfer(recipient, withdrawable),
            "API3 token transfer failed"
        );
        emit Withdrawn(recipient, withdrawable);
    }
```
After the require statement none of the statements are covered according to istanbul, seems strange because the tests do attempt to withdraw when tokens are available.

bbenligiray replied:
That's because the coverage package fails if `evm_setNextBlockTimestamp` is used in a test so I was having it skip those tests while measuring coverage:
https://github.com/api3dao/api3-contracts/blob/master/packages/timelock-manager/.solcover.js#L13
The latest version of hardhat / solidity-coverage doesn't have the same limitation though, the dependencies of these packages are way too old.

Ashar2shahid commented on packages/timelock-manager/contracts/TimeLockManagerReversible.sol:
```diff
@@ -0,0 +1,260 @@
+    /// @param recipient Original recipient of tokens
+    /// @param destination Destination of the tokens locked by the reverted
+    /// timelock
+    function revertTimelock(address recipient, address destination)
```
got it

Ashar2shahid commented on packages/timelock-manager/contracts/TimeLockManagerReversible.sol:
```diff
@@ -0,0 +1,260 @@
+    /// a recipient
+    /// @param recipient Original recipient of tokens
+    /// @param destination Destination of the excess tokens vested to the addresss
+    function stopVesting(address recipient, address destination)
```
my solidity formatter in vs code, ended up doing this, will disable it going forward.

bbenligiray replied:
Which one is that? We can pick one and agree on a common configuration.

Ashar2shahid replied:
I am actually using the solidity plugin in vscode that comes with solhint linter and prettier plugin for code formatting. [Link to the plugin](https://github.com/juanfranblanco/vscode-solidity#formatting-using-prettier-and-the-prettier-solidity-plugin). I am not sure if it would be possible to enforce one line per argument in the formatter, that's why I disabled it and manually reformatted it.

bbenligiray replied:
Created an issue
https://github.com/api3dao/airnode/issues/247

Ashar2shahid commented on packages/helpers/src/deployer.js:
```diff
@@ -19,6 +19,22 @@ async function deployTimelockManager(deployer, ownerAddress, api3TokenAddress) {
   return timelockManager;
 }
 
+async function deployTimelockManagerReversible(
```
I just added the deployer script, you can have a look. Its using the DAO address for the owner, since there wasn't any other address available in the deployments.json that I think was appropriate.

Ashar2shahid commented on packages/timelock-manager/deploy/2-deploy-timelock-manger-reversible.js:
```diff
@@ -0,0 +1,20 @@
+  const { deploy, log } = deployments;
+  const { deployer } = await getNamedAccounts();
+
+  const api3DaoVaultAddress =
```
Probably have to use a different address, couldn't figure out which one

bbenligiray replied:
Here
https://github.com/api3dao/api3-contracts/blob/master/packages/helpers/src/deploymentAddresses.json
Add a `timelockManagerReversibleOwner`, leave the value empty for `1` and `4`. Import it from there here and use that. We can populate it manually after we deploy the owner multisig.

bbenligiray commented on packages/timelock-manager/contracts/TimeLockManagerReversible.sol:
```diff
@@ -0,0 +1,255 @@
+            releaseEnd > releaseStart,
+            "releaseEnd not larger than releaseStart"
+        );
+        require(releaseStart > now, "releaseStart not in the future");
```
Can you remove this require and update the test accordingly please? This will give us a bit more flexibility regarding how we can use the contract.

bbenligiray reviewed (changes requested):
Thank you, both the implementation and tests look good (though we will have this reviewed by others before using).
In addition to the comments, can you rename the file `TimelockManagerReversible.sol`/`TimelockManagerReversible.sol.js` for consistency?

Ashar2shahid commented:
Addressed the issues above,  
>  can you rename the file TimelockManagerReversible.sol/TimelockManagerReversible.sol.js   

Are you referring to the spelling error in TimelockManagerReversible.sol.js or do you want me to completely rename both of these files, I have addressed the former in the commit
